core/src/main/java/hudson/model/UpdateSite.java

```diff
@@ -646,6 +674,246 @@ public Api getApi() {
+                List<WarningVersionRange> ranges = new ArrayList<>();
+                JSONArray versions = o.getJSONArray("versions");
+                for (int i = 0; i < versions.size(); i++) {
+                    WarningVersionRange range = new WarningVersionRange(versions.getJSONObject(i));
+                    ranges.add(range);
+                }
+                this.versionRanges = Collections.unmodifiableList(ranges);
+            } else {
+                this.versionRanges = Collections.emptyList();
+            }
+        }
+
+        /**
+         * Two objects are considered equal if they are the same type and have the same ID.
+         *
+         * @param o the other object
+         * @return true iff this object and the argument are considered equal
+         */
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (!(o instanceof Warning)) return false;
+
+            Warning warning = (Warning) o;
+
+            return id.equals(warning.id);
+        }
+
+        @Override
+        public int hashCode() {
+            return id.hashCode();
+        }
+
+        /**
+         * Returns true iff this warning is for Jenkins core
+         * @return true iff this warning is for Jenkins core
+         */
+        public boolean isCoreWarning() {
+            return this.type == Type.CORE;
+        }
+
+        /**
+         * Returns true iff this warning is for a plugin
+         * @return true iff this warning is for a plugin
+         */
+        public boolean isPluginWarning() {
+            return this.type == Type.PLUGIN;
+        }
+
+        public boolean isPluginWarning(@Nonnull String pluginName) {
+            return isPluginWarning() && pluginName.equals(this.component);
+        }
+
+        /**
+         * Returns true if this warning is relevant to the current configuration
+         * @return true if this warning is relevant to the current configuration
+         */
+        public boolean isRelevant() {
+            switch (this.type) {
+                case CORE:
+                    VersionNumber current = Jenkins.getVersion();
+
+                    if (!isRelevantToVersion(current)) {
+                        return false;
+                    }
+                    return true;
+                case PLUGIN:
+
+                    // check whether plugin is installed
+                    PluginWrapper plugin = Jenkins.getInstance().getPluginManager().getPlugin(this.component);
+                    if (plugin == null) {
+                        return false;
+                    }
+
+                    // check whether warning is relevant to installed version
+                    VersionNumber currentCore = plugin.getVersionNumber();
+                    if (!isRelevantToVersion(currentCore)) {
+                        return false;
+                    }
+                    return true;
+                case UNKNOWN:
```

Comment:
I would document `UNKNOWN` to be a placeholder for type Jenkins can not understand but should never appear in the metadata (I guess that is the intention).

Btw, this switch statement can be replaced by polymorphism with enum fields providing specific implementations.